src/lib_json/json_reader.cpp

```diff
@@ -1334,12 +1347,20 @@ void OurReader::addComment(Location begin, Location end,
   }
 }
 
-bool OurReader::readCStyleComment() {
+bool OurReader::readCStyleComment(bool* containsNewLineResult) {
+  if (containsNewLineResult) {
+    *containsNewLineResult = false;
+  }
+
   while ((current_ + 1) < end_) {
     Char c = getNextChar();
-    if (c == '*' && *current_ == '/')
+    if (c == '*' && *current_ == '/') {
       break;
+    } else if (containsNewLineResult && (c == '\r' || c == '\n')) {
```

Comment:
In general, we want round-trips to be possible. One problem with comments is that they are hard to reproduce exactly as they are. That's just life.

However, I'm not sure we should worry about `\r` for comments at all. If the tests pass without this, i.e. if we don't expect to break anything, I'd favor dropping the `\r` here. But this is not a strong objection.

Reply:
Sounds good, I'm happy to drop the carriage return.

The speed up comes from  the fact that if we have multiple comments after a value, we keep checking whether they belong to the value. E.g. see the following:

foo: bar, /*C1: bar is an important thingymajig*//*C2: this sectionexplainsthenext complicated thingy with a long verbose section*//*C3: Here are some todos for this section*/\nbaz: 1, ....

The second and third comments obviously don't belong to foo: bar, but currently we check everything after bar for a newline to see if C2 and C3 belong to foo:bar. In the below fuzzer test this totally wrecks our performance.

`    {"": /* /00"±ª":4/**/98"¬ÿ":8/**/99"¬":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":;{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"":9/**/97"¬ÿ":8/*Î/98"¬ÿ":8{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"ÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌ¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/*""*/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*":8/**/99"3":8/*ÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/98"¬ÿ":8/**/99"¬":8/**/9"9ÿ":{"":9/*":8/**":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/*[:[][[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬]":8/**/99"¬":8/**/9"¬[ÿÿA/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":9/**/99"¬":8/**/9"¬[ÿÿ{"":9/**//9"¬[ÿÿÿ/":{"":8/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/*[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[¥¤¤**/98"¬[":{"":9/**/92"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/9"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"":9/**/97"¬ÿ":8/*Ï/4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/***/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":;{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"":9/**/97"¬ÿ":8/*Î/98"¬ÿ":8{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/*""*/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*":8/**/99"3":8/*ÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/98"¬ÿ":8/**/99"¬":8/**/9"9ÿ":{"":9/*":8/**":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/*[:[][[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[*/99"¬":8stackLimit/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬]":8/**/99"¬":8/**/9"¬[ÿÿA/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":9/**/99"¬":8/**/9"¬[ÿÿ{"":9/**//9"¬[ÿÿÿ/":{"":8/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/*[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[¥¤¤**/98"¬[":{"":9/**/92"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"+¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/9"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"":9/**/97"¬ÿ":8/*Ï/4/**/9"¬ÿ":8/**/99"¬":8**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©©/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"":9/**/97"¬ÿ":8/*Î/98"¬ÿ":8{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/*""*/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*":8/**/99"3":8/*ÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/98"¬ÿ":8/**/99"¬":8/**/9"9ÿ":{"":9/*":8/**":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/*[:[][[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬]":8/**/99"¬":8/**/9"¬[ÿÿA/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":9/**/99"¬":8/**/9"¬[ÿÿ{"":9/**//9"¬[ÿÿÿ/":{"":8/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/*[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[¥¤¤**/98"¬[":{"":9/**/92"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"+¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/9"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"¬":8/**/9"¬[ÿÿA/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/*":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"3":8/**/9"4¬ÿÿ":8/**/9"9ÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":9/**/99"¬":8/**/9"¬[ÿÿ{"":9/**//9"¬[ÿÿÿ/":{"":8/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/*[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[¥¤¤**/98"¬[":{"":9/**/92"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"+¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**//9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/**/¬"99ª":{"":9/**/98"¬ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/9"ÿ":8/**/99"¬":8/***/98"¬/ÿ":{"":9/**/97"¬ÿ":8/*Ï/4/**/9"¬ÿ":8/**/99"¬":8**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"ÿ":8/**/99"¬":8/***/98"¬[":{"":9/**/99"¬ÿ":8/**/9"ª":4/**/98"¬":8/**/9"9ÿ":{"":9/*/99"¬":8/**/9"¬[ÿÿÿ/":{"":4/**/9"¬ÿ":8/**/99"¬":8/**/9"¬[ÿÿÿ/":{"":9/**/9"¬":8/*`